frontend/packages/console-dynamic-plugin-sdk/src/extensions/user-preferences.ts

```diff
@@ -0,0 +1,85 @@
+import * as React from 'react';
+import { JSONSchema7Type } from 'json-schema';
+import { Extension, ExtensionDeclaration, CodeRef } from '../types';
+
+export enum UserPreferenceFieldType {
+  dropdown = 'dropdown',
+  checkbox = 'checkbox',
+  custom = 'custom',
+}
+
+export type UserPreferenceDropdownField = {
+  type: UserPreferenceFieldType.dropdown;
+  userSettingsKey: string;
+  defaultValue?: string;
+  options: {
+    value: string;
+    label: string;
+  }[];
+};
+
+export type UserPreferenceCheckboxFieldValue = string | number | boolean;
+
+export type UserPreferenceCheckboxField = {
+  type: UserPreferenceFieldType.checkbox;
+  userSettingsKey: string;
+  label: string;
+  trueValue: UserPreferenceCheckboxFieldValue;
+  falseValue: UserPreferenceCheckboxFieldValue;
+  defaultValue?: UserPreferenceCheckboxFieldValue;
+};
+
+export type UserPreferenceCustomField = {
+  type: UserPreferenceFieldType.custom;
+  component: CodeRef<React.ComponentType>;
+  props?: { [key: string]: JSONSchema7Type };
+};
+
+export type UserPreferenceField =
+  | UserPreferenceDropdownField
+  | UserPreferenceCheckboxField
+  | UserPreferenceCustomField;
+
+export type UserPreferenceGroup = ExtensionDeclaration<
+  'console.user-preference/group',
+  {
+    /** ID used to identify the user preference group. */
+    id: string;
+    /** The label of the user preference group */
+    label: string;
+    /** ID of user preference group before which this group should be placed */
+    insertBefore?: string;
+    /** ID of user preference group after which this group should be placed */
+    insertAfter?: string;
+  }
+>;
+
+export type UserPreferenceItem = ExtensionDeclaration<
+  'console.user-preference/item',
+  {
+    /** ID used to identify the user preference item. The value for ID is generally the key for the user preference in config map */
+    id: string;
```

Comment:
Didn't we say that the preference key in the configmap for the pre-defined primitive fields would actually be a property named `key` on the `field` itself and not the `id`? 
This way we'd separate out the two concerns and only use the ID for ordering?
cc @jerolimov

Reply:
That's right. The comment is wrong. We have a `userSettingsKey` defined in the field attribute.

In `UserPreferenceDropdownField` and `UserPreferenceCheckboxField` which could maybe be unified, but for me that's not so important at the moment. But Christian is right we should change the comment here:

```suggestion
    /** ID used to identify the user preference item, referenced in insertAfter and insertBefore to define the item order. */
    id: string;
```